### PR TITLE
[Fix] Immutable MetaBundle stored in cache.Cache

### PR DESCRIPTION
**Pull request by clamoriniere**
In order to avoid race-condition with data stored in cache, the object
stored in the cache should never be modify after insertion.

Also add Copy() methods helper.

### CODE REVIEW

File: pkg/util/kubernetes/apiserver/store.go [hunk 1]
```diff
@@ -55,27 +55,24 @@ func (m *metaBundleStore) get(nodeName string) (*metadataMapperBundle, bool) {
...
+func (m *metaBundleStore) getCopyOrNew(nodeName string) *metadataMapperBundle {
 	cacheKey := agentcache.BuildAgentKey(metadataMapperCachePrefix, nodeName)
 
-	var metaBundle *metadataMapperBundle
+	metaBundle := newMetadataMapperBundle()
 
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
 	v, ok := m.cache.Get(cacheKey)
```
Comment (mfpierre) [inline pkg/util/kubernetes/apiserver/store.go]: where are we setting the cache?

File: pkg/clusteragent/api/v1/types.go [hunk 2]
```diff
@@ -45,6 +45,24 @@ func NewNamespacesPodsStringsSet() NamespacesPodsStringsSet {
 	return make(NamespacesPodsStringsSet)
 }
 
+// Copy used to copy NamespacesPodsStringsSet in another NamespacesPodsStringsSet
+func (m NamespacesPodsStringsSet) Copy(old *NamespacesPodsStringsSet) NamespacesPodsStringsSet {
```
Comment (mfpierre) [inline pkg/clusteragent/api/v1/types.go]: Should probably be unit tested

File: pkg/util/kubernetes/apiserver/metadata_controller.go [hunk 3]
```diff
@@ -245,16 +246,13 @@ func (m *MetadataController) mapEndpoints(endpoints *corev1.Endpoints) error {
 	svc := endpoints.Name
 	namespace := endpoints.Namespace
 	for nodeName, ns := range nodeToPods {
-		metaBundle := m.store.getOrCreate(nodeName)
-
-		metaBundle.m.Lock()
+		metaBundle := m.store.getCopyOrNew(nodeName)
 		metaBundle.Services.Delete(namespace, svc) // cleanup pods deleted from the service
```
Comment (mfpierre) [inline pkg/util/kubernetes/apiserver/metadata_controller.go]: this won't affect the one in the store though ? 🤔

Comment (clamoriniere) [reply pkg/util/kubernetes/apiserver/metadata_controller.go]: it does, we do a `store.set(...)` https://github.com/DataDog/datadog-agent/pull/3404/files/2c04b9d3176a63bef33ed50508ca5dd948bcacf6#diff-88ac0183412502946693e4490295addaL257

File: pkg/util/kubernetes/apiserver/store.go (hunk 1 above)
Comment (clamoriniere) [reply pkg/util/kubernetes/apiserver/store.go]: we set in the cache outside this method. I change the name in order to reflect the change.
I also check every where this method was call, in order to be sure that it works as before

File: pkg/clusteragent/api/v1/types.go (hunk 2 above)
Comment (clamoriniere) [reply pkg/clusteragent/api/v1/types.go]: unit-test added

**Review by mfpierre — APPROVED** [review]
OK to try this fix, but not a fan of assuming the consumer will work with copies, this can be error prone in the future, let's add some backlog to rework the store/cache logic

**Comment by clamoriniere** [comment]
@mfpierre I agree with you. This store/cache logic needs a bit more rework.
The current implementation allow a consumer to work with the current cached object by using: `store.get()`, `store.getCopyOrNew()` is only use when the consumer know that it needs to modify the object.

**Pull request merged**